### PR TITLE
AMP-Carousel with AMP-List

### PR DESCRIPTION
**Pull request by bthomas16**
Adding an advanced component and endpoint so potential users can learn how a dynamic carousel might look and work

### CODE REVIEW

**Comment by bthomas16** [comment]
I signed it!

File: src/json/amp-carousel_with_amp-list.json [hunk 1]
```diff
@@ -0,0 +1,82 @@
+{
```
Comment (sebastianbenz) [inline src/json/amp-carousel_with_amp-list.json]: Can you please use some of our own hosted images, for example: `/json/related_products.json` should work quite well in this case.

File: src/30_Advanced/Amp-Carousel_with_Amp-List.html [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+<html ⚡>
+ <head>
+   <meta charset="utf-8">
+   <link rel="canonical" href="hello-world.html">
+   <meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1">
+   <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
+   <script async src="https://cdn.ampproject.org/v0.js"></script>
+   <!-- #### Setup -->
+  <!--
+    Import the `amp-list` and `amp-mustache` and `amp-carousel` and `amp-bind` components.
```
Comment (sebastianbenz) [inline src/30_Advanced/Amp-Carousel_with_Amp-List.html]: Can you split the import comment so that each extension import has a comment (this way all imports will be shown in the generated doc page). Similar to here: https://github.com/ampproject/amp-by-example/blob/master/src/40_Dynamic_AMP/Multi_Page_Flow.html#L22

File: src/30_Advanced/Amp-Carousel_with_Amp-List.html [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            }
+        </script>
+    </amp-state>
+
+    <!-- This is a simplified version of our production release, 
+         some code is missing, but concepts will be explained -->
+    <div class="carousel-wrapper" >
+        <amp-list 
+            width="325" height="325" 
+            layout="fill"
```
Comment (sebastianbenz) [inline src/30_Advanced/Amp-Carousel_with_Amp-List.html]: suggestion: use a fixed-height layout which uses the same height as the carousel. This way you shouldn't need additional css.

File: src/30_Advanced/Amp-Carousel_with_Amp-List.html [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        <amp-list 
+            width="325" height="325" 
+            layout="fill" 
+            src="/amp-carousel_with_amp-list.json" >
+            <template type="amp-mustache">
+                <amp-carousel id="alternate-images-carousel" width="325" height="325" 
+                    type="slides" 
+                    layout="responsive" 
+                    controls
+                    on="slideChange:AMP.setState({selected: { slide: event.index }})">
```
Comment (sebastianbenz) [inline src/30_Advanced/Amp-Carousel_with_Amp-List.html]: I don't think this line is needed.

File: src/30_Advanced/Amp-Carousel_with_Amp-List.html [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+                    <!-- in production, this primary image is set from the modelId,
```
[the remark this answers is not in the document]
Comment (sebastianbenz) [reply src/30_Advanced/Amp-Carousel_with_Amp-List.html]: Can you maybe phrase the text more like a documentation (and don't mention your own prod implementation which is hard to follow if someone doesn't know it).

File: src/30_Advanced/Amp-Carousel_with_Amp-List.html [hunk 6]
```diff
@@ -0,0 +1,94 @@
+<!---{
+  "preview": "default"
+  }--->
+<!--
+  #### Introduction
+  AMP users might want to pull data from an endpoint to display in a carousel. This example incoporates a dynamic amp-carousel as well as showing how to not only rely on the images served from the endpoint
```
Comment (sebastianbenz) [inline src/30_Advanced/Amp-Carousel_with_Amp-List.html]: Can you explain this part: 'as well as showing how to not only rely on the images served from the endpoint' a bit more? How relevant is it to the sample?

Comment (bthomas16) [reply src/30_Advanced/Amp-Carousel_with_Amp-List.html]: We are using AMP with C# for e-commerce, so the way we are using it:

src="<%= Model.Media.PrimaryHeroImage %>  [src]="skuState[skuState.currentSkuState].HeroImage". 

Depending on if a user has selected a different color/sku, we want that image to be the first image in our carousel.

File: src/30_Advanced/Amp-Carousel_with_Amp-List.html [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+   <script async src="https://cdn.ampproject.org/v0.js"></script>
+   <!-- #### Setup -->
+  <!--
+    Import the `amp-list` and `amp-mustache` and `amp-carousel` and `amp-bind` components.
+  -->
+  <script async custom-element="amp-list" src="https://cdn.ampproject.org/v0/amp-list-0.1.js"></script>
+  <script async custom-template="amp-mustache" src="https://cdn.ampproject.org/v0/amp-mustache-0.2.js"></script>
+  <script async custom-element="amp-carousel" src="https://cdn.ampproject.org/v0/amp-carousel-0.1.js"></script>
+  <script async custom-element="amp-bind" src="https://cdn.ampproject.org/v0/amp-bind-0.1.js"></script>
```
[the remark this answers is not in the document]
Comment (bthomas16) [reply src/30_Advanced/Amp-Carousel_with_Amp-List.html]: Sorry, got too quick cutting some of our code off. Our production version uses 'dots' under the carousel to indicate which slide you're currently viewing. We do not need amp bind for this simplified example

File: src/30_Advanced/Amp-Carousel_with_Amp-List.html [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    }
+    .hero-carousel-image {
+        max-width:225px;
+        height:auto;
+        width:100%;
+    }
+   </style>
+ </head>
+ <body>
+    <amp-state id="selected">
```
[the remark this answers is not in the document]
Comment (bthomas16) [reply src/30_Advanced/Amp-Carousel_with_Amp-List.html]: I will delete this from my next commit

File: src/30_Advanced/Amp-Carousel_with_Amp-List.html [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+  -->
+  <script async custom-element="amp-list" src="https://cdn.ampproject.org/v0/amp-list-0.1.js"></script>
+  <script async custom-template="amp-mustache" src="https://cdn.ampproject.org/v0/amp-mustache-0.2.js"></script>
+  <script async custom-element="amp-carousel" src="https://cdn.ampproject.org/v0/amp-carousel-0.1.js"></script>
+  <script async custom-element="amp-bind" src="https://cdn.ampproject.org/v0/amp-bind-0.1.js"></script>
+   <!--
+    #### CSS
+  -->
+   <style amp-custom>
+       #alternate-images-carousel {
```
[the remark this answers is not in the document]
Comment (bthomas16) [reply src/30_Advanced/Amp-Carousel_with_Amp-List.html]: I will delete this from my next commit

File: src/30_Advanced/Amp-Carousel_with_Amp-List.html (hunk 5 above)
[the remark this answers is not in the document]
Comment (bthomas16) [reply src/30_Advanced/Amp-Carousel_with_Amp-List.html]: I will fix this in my next commit

File: src/json/amp-carousel_with_amp-list.json (hunk 1 above)
Comment (bthomas16) [reply src/json/amp-carousel_with_amp-list.json]: I will edit the api to point here and the mustache template to reference it properly

File: src/30_Advanced/Amp-Carousel_with_Amp-List.html (hunk 2 above)
Comment (bthomas16) [reply src/30_Advanced/Amp-Carousel_with_Amp-List.html]: Will fix this in my next commit

File: src/30_Advanced/Amp-Carousel_with_Amp-List.html (hunk 3 above)
Comment (bthomas16) [reply src/30_Advanced/Amp-Carousel_with_Amp-List.html]: Will play with this and verify still works before my next commit

File: src/30_Advanced/Amp-Carousel_with_Amp-List.html (hunk 4 above)
Comment (bthomas16) [reply src/30_Advanced/Amp-Carousel_with_Amp-List.html]: Sorry, this if for the 'dots' under the carousel

File: src/30_Advanced/Amp-Carousel_with_Amp-List.html [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+ <body>
+    <amp-state id="selected">
+        <script type="application/json">
+            {
+            "slide": 0
+            }
+        </script>
+    </amp-state>
+
+    <!-- This is a simplified version of our production release,
```
[the remark this answers is not in the document]
Comment (bthomas16) [reply src/30_Advanced/Amp-Carousel_with_Amp-List.html]: Will add this to my next commit
